Add table block column alignment

talldan opened the pull request:
## Description
Adds column alignment options to the table block. Closes #15823

**Dev decisions**
Code-wise, the part where I've had to break from the norm is by using a `data-align` attribute to store the alignment value for a cell. This is because the cell attributes are deeply nested and sourced using a `query` matcher, so normal attributes can't be used.

Applying an alignment adds a class name and `data-align` attribute to each table cell in a column. There aren't many alternatives to this approach that I can think of, apart from using inline styles. `colgroup`/`col` elements don't support text align, so they're out of the question.

## How has this been tested?
Added unit and e2e tests.

Manual testing steps:
1. Add a table block to a post
2. Add some text to the first column of the table.
3. Ensure one of the cells in the first column is selected
4. From the toolbar try each of the column alignment options
5. Verify that text in the each cell in the column is aligned in the editor
6. Preview the post
7. Verify that the text is aligned when viewing the post.

## Screenshots <!-- if applicable -->
**Editor**
<img width="425" alt="Screen Shot 2019-07-30 at 10 46 50 am" src="https://user-images.githubusercontent.com/677833/62096810-7a2c6580-b2b7-11e9-9ed7-daa666eb7f99.png">

**Post**
<img width="466" alt="Screen Shot 2019-07-30 at 10 47 45 am" src="https://user-images.githubusercontent.com/677833/62096824-903a2600-b2b7-11e9-8b15-dac05e77d81b.png">

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
New feature (non-breaking change which adds functionality)
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://wordpress.org/gutenberg/handbook/designers-developers/ -->

talldan committed:
Initial step at adding alignment options in table block edit component

Fix block.json attributes

Add cell alignment to save function

Add tests for table cell alignment

Tidy up test comments

Update JSDocs for new functions

Add unit tests

Use classnames for cell alignment

Use shorthand property

Remove alignment styles in favour of global styles

Update copy of alignment buttons

Rework cell selection border to use an actual border around a pseudoelement instead of box-shadow

mapk commented:
I just tested this and it works wonderfully! Thanks, @talldan! 👍 from a design perspective. :shipit: 

<img width="638" alt="Screen Shot 2019-06-15 at 7 18 06 AM" src="https://user-images.githubusercontent.com/617986/59552612-1b48a200-8f3e-11e9-91a3-c2513f1e18ec.png">

talldan commented:
Thanks! I think I'll switch this to using classnames as #16035 does, as that's now been merged.

mtias commented:
Should this apply to the whole table, per column, or on individual cells?

mapk commented:
> Should this apply to the whole table, per column, or on individual cells?

Great question, @mtias. Right now the block only allows selection of a single cell at a time. I would love to see selections happen across rows and columns. Once this works, we can look at alignment across these.

talldan commented:
I've refactored this to use classes, so it's ready for a code review 👍

youknowriad commented on packages/block-library/src/table/block.json:
```diff
@@ -29,6 +29,11 @@
+						"align": {
+							"type": "string",
+							"source": "attribute",
+							"attribute": "data-align"
```
Interesting. since it applied as a class, makes me wonder if at some point, we'd want "classNames" sources. That said, it can be complex and I know it's a deliberate decision to restrict the number of sources available. css @aduth

aduth replied:
Yeah, I think we'd considered once before a source serialized as the presence of a class name. Personally I don't have much of an issue with the `data-` attributes; they can be styled just as well, are a bit easier to parse, and fit well as serving the purpose of a data attribute.

>Custom data attributes are intended to store custom data, state, annotations, and similar, private to the page or application, for which there are no more appropriate attributes or elements.

https://html.spec.whatwg.org/multipage/dom.html#embedding-custom-non-visible-data-with-the-data-*-attributes

That being said, I'm not really sure why we bother to serialize this in the HTML at all, if we're not using it for anything like styling? Why not just embed it within the comment delimiters (i.e. omit `source`)?

talldan replied:
@aduth As this attribute is in a `query`, my understanding is that `source` can't be omitted. A `classNames` source could be useful in this use case.

An argument against `data-` attributes is that it's double storing data, but then it's not too much of a problem with the reactive nature of blocks, it's not like there's extra business logic to keep them in sync.

ellatrix replied:
Why use the class at all? Can we not just use the data attribute?

talldan replied:
There are some global styles for text alignment:
https://github.com/WordPress/gutenberg/blob/4ac7dc9d2d93a1b3ab81b44a6fbd89952a44915e/packages/block-library/src/style.scss#L150-L161

Seems better to be able to use those and not have extra styles in the table block.

youknowriad commented on packages/block-library/src/table/edit.js:
```diff
@@ -465,6 +502,11 @@ export class TableEdit extends Component {
+					<AlignmentToolbar
+						value={ this.getCellAlignment() }
+						onChange={ ( nextAlign ) => this.onChangeCellAlignment( nextAlign ) }
+						isCollapsed={ true }
```
Nit: `={ true }` can be dropped.

talldan replied:
👍

youknowriad commented on packages/block-library/src/table/style.scss:
```diff
@@ -76,6 +76,18 @@
 		td {
 			border-color: transparent;
+
+			&.has-text-align-left {
```
Do we really need these classNames to be defined here? I thought these were meant to be global classNames?

talldan replied:
Well spotted, I didn't realise there were global styles. Removed those styles.

youknowriad reviewed (commented):
This seesm to be in a good shape. I left some minor comments.

ellatrix reviewed (changes requested):
I have some strong opinion about the user experience of this, sorry.

1. We should only allow table cells to be left- or right-aligned. The button can be a simple toggle. Center alignment in tables doesn't make so much sense. What kind of data will you align like that? The general rule of thumb is that textual data should be left-aligned, and numerical data should be right-aligned.
2. Another table "rule" is that each column should have the same alignment. This not only looks better,  it easier for the user to align the entire column at once.

mtias commented:
I think center alignment can be perfectly fine in some cases, and we should not dictate that. However, I think that single cell alignment is a bit excessive. I'd rather have a text alignment toolbar button that applies to the whole table at once (since columns is a bit more complicated).

ellatrix commented:
We can always add center alignment if it's much requested. It's easy to add, but hard to take away. Center alignment only makes tables much harder to scan.

Why is applying it to the column complicated? Aligning an entire table seems really strange to me. A table usually doesn't contain only numerical data (at the very least, the row headings won't be numerical).

talldan commented:
I'm not sure I see much difference between other formatting options and alignment. For example, I'm unlikely to want to apply `bold` formatting to just a single cell, instead I'd be more likely to select the entire header or footer row and choose bold formatting. However, this is the way it works in the table block today.

I started looking at column/row/table selection in another PR (https://github.com/WordPress/gutenberg/pull/16493). It's in a very early state, needs some design and accessibility work. Also needs some more work to make it actually work. I think it's something that'll be required as we start to add more options to the table block that can be applied to an individual or multiple cells. It would also be a good first stepping stone to more freeform multiple selection of cells.

I think without that, clicking align and having it apply to an entire column or table when there's only a single cell selected feels like an unexpected outcome. It'd be better to put the effort into #16493 in my opinion.

talldan commented:
Have rebased and updated this based the code review feedback (thanks for the review!).

I've added the 'Needs Decision' label, as a decision will have to be made on the how alignment is applied (per cell, column or table). The debate on that is in the four posts above this one for anyone reading.

talldan commented:
I've made a branch [`try/table-multi-selection-with-alignments`](https://github.com/WordPress/gutenberg/compare/try/table-multi-selection-with-alignments?expand=1) that merges this PR with #16493. Here's a demo:

![table-alignment](https://user-images.githubusercontent.com/677833/61207521-e5a6fc80-a727-11e9-9fd1-c27333e20df1.gif)

Pretty easy to modify any cell attributes across selections with this code. The tough bit is going to be applying rich text formatting across cells. I'm not entirely sure how to achieve that. Any ideas @ellatrix?

mapk commented:
> We should only allow table cells to be left- or right-aligned. The button can be a simple toggle. Center alignment in tables doesn't make so much sense. What kind of data will you align like that? 

@ellatrix, there are a couple examples that come to mind. Some of which can be seen in this graphic from wordpress.com/pricing below:

![Screen Shot 2019-07-24 at 9 44 43 AM](https://user-images.githubusercontent.com/617986/61812228-edd1fc80-adf7-11e9-926b-6a21c421d7a0.png)

1. I look forward to a time when we can add images/icons to the Table block. In which case, center-aligning checkboxes like this makes sense.
2. Center aligning the column headers also makes sense depending on the design.

> Another table "rule" is that each column should have the same alignment. This not only looks better, it easier for the user to align the entire column at once.

I love this and agree that moving these alignments to column-wide settings would be great!

@talldan while I like the per-cell adjustements I can make, I do long for the column-based alignments and will test that other PR today. One thing I noted is in @jasmussen's PR https://github.com/WordPress/gutenberg/pull/16557#issuecomment-514639436 he uses an arrow icon to indicate the dropdown of block alignments. Should we do the same here for the text alignments?

talldan commented:
@mapk I think we'll need to solve that a cell is shown selected when aligning, because that'll be confusing. I can look into highlighting the column that will be aligned when the alignment icon button is hovered, and also change the label text to 'Align column ...'. 

I think that would make it clear enough.

talldan committed:
Make cell alignment work across columns instead of individual cells

Add jsdocs

Inherit the alignment property of the first cell in the column when adding new rows

Update e2e tests

Use consistent letter casing

talldan commented:
I've updated this PR so that it now aligns entire columns. 

Tests are all green, so it should be ready for review.

ellatrix commented on packages/block-library/src/table/editor.scss:
```diff
@@ -41,9 +41,19 @@
-		border-color: $blue-medium-500;
-		box-shadow: inset 0 0 0 1px $blue-medium-500;
-		border-style: double;
+		position: relative;
```
What is the reason behind these changes? What does it have to do with alignment?

talldan replied:
It is a better way to apply a cell border than box-shadow, but not needed for this PR so I've removed these changes.

ellatrix commented on packages/block-library/src/table/state.js:
```diff
@@ -26,105 +28,186 @@ export function createTable( {
+ *
+ * @return {boolean} True if the cell is selected, false otherwise.
+ */
+export function isCellSelected( cellLocation, selection ) {
```
Why not have `isCellSelected` and `isColumnSelected`?

talldan replied:
`isColumnSelected` would only be used in one place, so it seems unnecessary at this stage to have two separate functions.

I'm exploring this more in #16493, where I do have separate functions for different types of selections as well as an `isCellSelected` for easily checking whether a given cell is in any type of selection:
https://github.com/WordPress/gutenberg/blob/ec1aa436ec4e4ce641799ec96e9bc04862405da9/packages/block-library/src/table/state.js#L221-L244

It'll be an easy refactor from this to that.

ellatrix replied:
That's fine, but then I wouldn't call it `isCellSelected`? That's misleading if you're checking if a column is selected. `isCellSelected( ..., 'column' )` doesn't make sense?

talldan replied:
Hmm, yeah, that's good feedback. I can work on a few naming improvements in future PRs as this functionality develops. Something like `isCellWithinSelection` might be more appropriate.

Thanks for the reviews btw!

ellatrix commented on packages/block-library/src/table/state.js:
```diff
@@ -26,105 +28,186 @@ export function createTable( {
- * @param {string} options.section  Section in which to insert the row.
- * @param {number} options.rowIndex Row index at which to insert the row.
+ * @param {Object} state                Current table state.
+ * @param {string} options.sectionName  Section in which to insert the row.
```
Looks like there's a space too much?

ellatrix commented on packages/block-library/src/table/state.js:
```diff
@@ -26,105 +28,186 @@ export function createTable( {
 } ) {
-	const cellCount = columnCount || state[ section ][ 0 ].cells.length;
+	const firstRow = getFirstRow( state );
+	const cellCount = columnCount || get( firstRow, [ 'cells', 'length' ], 2 );
```
Why does it default to 2?

talldan replied:
That's a fair comment. I thought `2` because that's what the table block defaults to when created.

Maybe this should instead be considered unexpected and the `insertRow` function should return early with no effect if `cellCount` is undefined.

talldan replied:
I changed it to return early with unchanged state if `cellCount` is unknown.

ellatrix replied:
I think so too. If there's no amount of rows, there's nothing to be done? If you'd like to keep a default, I feel like it shouldn't be hardcoded here, but set by the component instead (which knows the default row count).

ellatrix commented on packages/block-library/src/table/state.js:
```diff
@@ -26,105 +28,186 @@ export function createTable( {
-				} ) ),
+				cells: times( cellCount, ( index ) => {
+					const firstCellInColumn = get( firstRow, [ 'cells', index ], {} );
+					const inheritedAttributes = pick( firstCellInColumn, INHERITED_COLUMN_ATTRIBUTES );
```
Why is alignment inherited?

talldan replied:
When adding a new row, the newly added cells will keep the same alignment as the rest of their adjacent column.

As already mentioned in this PR, it seems like all cells in a column should have the same alignment, so this maintains that experience for the user. In my testing it felt like the right way to go.

ellatrix replied:
Sounds good

ellatrix commented on packages/block-library/src/table/state.js:
```diff
@@ -26,105 +28,186 @@ export function createTable( {
 	rowIndex,
 	columnIndex,
-	content,
+	attributeName,
```
I think this would read better if `attributeName` is a separate argument. E.g. `getCellAttribute( state, 'align', cell )`.

talldan replied:
Agreed it could be improved. I've changed it to `getCellAttribute( state, cell, 'align' )`.

ellatrix replied:
Great! It seems like a good idea to keep `cellLocation` as a separate entity. Could this object be documented somewhere to the type can be used in the docs of all these functions?

ellatrix commented on packages/block-library/src/table/edit.js:
```diff
@@ -156,14 +179,73 @@ export class TableEdit extends Component {
+		setAttributes( updateSelectedCell(
+			attributes,
+			selectedCell,
+			( cellAttributes ) => ( { ...cellAttributes, content } ),
```
Why do we need a callback here? Can this not be done directly by `updateSelectedCell`?

talldan replied:
It might be overkill, but I feel it's a more useful API. Being able to read the cell's attributes provides an opportunity for doing things like toggling values or incrementing/decrementing values when updating them.

talldan committed:
Address review feedback

Update params of getCellAttribute function

Do nothing if insertRows is unable to determine the correct cellCount for the insertion

paaljoachim commented:
I assume one can then multi select cells in a column and have all of them align in one go.

talldan commented:
> I assume one can then multi select cells in a column and have all of them align in one go.

@paaljoachim Sort of, the table still only has column-based selection, but alignments apply to the column that the selected cell is in.

ellatrix commented on packages/e2e-tests/specs/blocks/table.test.js:
```diff
@@ -10,6 +15,17 @@ import {
+ */
+async function changeCellAlignment( align ) {
+	await clickBlockToolbarButton( 'Change column alignment' );
+	const alignButton = await page.$x( `//button[text()='Align Column ${ capitalize( align ) }']` );
```
Can the `clickButton` utility function be used here?

talldan replied:
Genuinely didn't know that existed. Will do a follow up PR to tidy up the tests.

ellatrix commented on packages/e2e-tests/specs/blocks/table.test.js:
```diff
@@ -154,4 +170,40 @@ describe( 'Table', () => {
+		await page.keyboard.type( '4' );
+
+		// Create the table.
+		const [ createButton ] = await page.$x( createButtonSelector );
```
Also here, could `clickButton` be used?
